### ISSUE TITLE
Why does the score of TPOT recommend classifier is lower than LinearSVC?

### ISSUE THREAD

**Issue by BenjiKCF** [rhiever/tpot#369]
So I have found out that LinearSVC is in TPOT classifier and I have been using it for my model and get a pretty decent score (0.95 in sklearn score). 

    def process(stock):
      df = format_data(stock)
      df[['HSI Volume', 'HSI', stock]] = df[['HSI Volume', 'HSI', stock]].pct_change()

    # shift future value to current date
      df[stock+'_future'] = df[stock].shift(-1)
      df.replace([-np.inf, np.inf], np.nan, inplace=True)
      df.dropna(inplace=True)
      df['class'] = list(map(create_labels, df[stock], df[stock+'_future']))
      X = np.array(df.drop(['class', stock+'_future'], 1)) # 1 = column
      # X = preprocessing.scale(X)
      y = np.array(df['class'])

      X_train, X_test, y_train, y_test = model_selection.train_test_split(X, y, test_size=0.2)

      tpot = TPOTClassifier(generations = 10, verbosity=2)
      fitting = tpot.fit(X_train, y_train)
      prediction = tpot.score(X_test, y_test)
      tpot.export('pipeline.py')
      return fitting, prediction

After ten generations: TPOT recommended GaussianNB and it scores about 0.77 in sklearn score. 

    Generation 1 - Current best internal CV score: 0.5322255571                     
    Generation 2 - Current best internal CV score: 0.55453535828                    
    Generation 3 - Current best internal CV score: 0.55453535828                    
    Generation 4 - Current best internal CV score: 0.55453535828                    
    Generation 5 - Current best internal CV score: 0.587469903893                   
    Generation 6 - Current best internal CV score: 0.587469903893                   
    Generation 7 - Current best internal CV score: 0.597194474469                   
    Generation 8 - Current best internal CV score: 0.597194474469                   
    Generation 9 - Current best internal CV score: 0.597194474469                   
    Generation 10 - Current best internal CV score: 0.597194474469                  
                                                                                    
    Best pipeline: GaussianNB(RBFSampler(input_matrix, 0.22))
    (None, 0.54637855142056824)

I am just curious that why LinearSVC scores higher but TPOT did not recommend. Is it because of the scoring mechanism is different and hence lead to a different optimal classifier?

Thank you so much!

**Comment by weixuanfu2016** [comment]
The default score found

Hmm, I am not sure what happened. Below is my guesses:

1. As you thought, the default scoring mechanism is different. It is called balanced accuracy which computes each class' accuracy on a per-class basis using a one-vs-rest encoding, then computes an unweighted average of the class accuracies. The default scoring function in `LinearSVC` is mean accuracy of self.predict(X) wrt. y. So if a few classes high classification error, it may cause this score very low. You may try add `scoring="accuracy"` in TPOTClassifier for using the same scoring function as `LinearSVC` 

2 . Maybe `LinearSVC` was not randomly generated in the founder population (generation 0) and not randomly generated by mutation in later generations. The default population size in TPOT is 100 and default mutation rate is 0.9. Maybe you could try to increase population size (with `population_size` parameter), like 500 or change a random seed with `random_state` parameter. (It is noted that current version of TPOT may not reproduce results with same random seed. This issue will be fixed in next version)

Please attach more details about your data (how many samples and features or raw data if available) and running environment.

**Comment by rhiever** [comment]
I suspect this happened because TPOT uses a different default scoring function than sklearn. TPOT uses balanced accuracy, whereas sklearn uses accuracy. In the upcoming 0.7 release, we have changed it so TPOT uses accuracy by default to avoid this kind of confusion.

**Comment by BenjiKCF** [comment]
Thank you so much weixuanfu201 and Rhiver. Both of your answers help me to understand more.

**Comment by BenjiKCF** [comment]
Thank you all, after running for 100generations, it releases the Best pipeline: GradientBoostingClassifier(GaussianNB(FeatureAgglomeration(GradientBoostingClassifier(PCA(StandardScaler(CombineDFs(input_matrix, input_matrix)), 8), 19.0, 4.0), 24, 9)), 0.01, 28.0)
(None, 0.94935064935064939) with the same accuracy scale.

**Comment by weixuanfu2016** [comment]
Great, good to know!
